### PR TITLE
Feature/domain setting

### PR DESCRIPTION
**Pull request by rejoshed**
Issue # -- CTECH-94

**This PR was rebased upon the Tilt fixup branch. Once that is merged, it will not show those changes.** 

Description of changes:
If specified, resources will be created under account and domain ids.

Here a VM can be seen under my TestAccountName account which resides in my TestDomain.
image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: api/v1alpha3/cloudstackcluster_webhook.go [hunk 1]
```diff
@@ -54,19 +55,20 @@ var _ webhook.Validator = &CloudStackCluster{}
...
-	)
+	var errorList field.ErrorList
 
 	// IdentityRefs must be Secrets.
-	if spec.IdentityRef != nil && spec.IdentityRef.Kind != defaultIdentityRefKind {
+	if r.Spec.IdentityRef != nil && r.Spec.IdentityRef.Kind != defaultIdentityRefKind {
 		errorList = append(errorList, field.Forbidden(field.NewPath("spec", "identityRef", "kind"), "must be a Secret"))
 	}
 
+	if (r.Spec.Account == "") != (r.Spec.Domain == "") {
```
Comment (rejoshed) [inline api/v1alpha3/cloudstackcluster_webhook.go:65]: Require both account and domain if set.

File: go.mod [hunk 2]
```diff
@@ -9,7 +9,7 @@ require (
 	github.com/golang/mock v1.6.0
 	github.com/hashicorp/go-multierror v1.1.1
 	github.com/onsi/ginkgo v1.16.4
-	github.com/onsi/gomega v1.16.0
+	github.com/onsi/gomega v1.17.0
```
Comment (rejoshed) [inline go.mod:12]: Bumped for better gomega matchers.

File: pkg/cloud/helpers_test.go [hunk 3]
```diff
@@ -46,3 +49,27 @@ func getConfigPath(filename string) string {
 	dir, _ := os.Getwd()
 	return path.Join(dir, FixturePath, filename)
 }
+
+// This matcher is used to make gomega matching compatible with gomock parameter matching.
+// It's pretty awesome!
```
Comment (rejoshed) [inline pkg/cloud/helpers_test.go:54]: Like seriously!!!

File: pkg/cloud/instance_test.go [hunk 4]
```diff
@@ -181,6 +182,49 @@ var _ = Describe("Instance", func() {
 			Ω(client.GetOrCreateVMInstance(csMachine, machine, csCluster, "")).Should(MatchError("unknown err"))
 		})
 
+		describeDomainAccountTest := func(desc string) func(string, string) string {
+			return func(account string, domainID string) string {
+				return fmt.Sprintf("%s and %s to beet to, %s ", account, domainID, desc)
```
Comment (jweite-amazon) [inline pkg/cloud/instance_test.go:187]: beet?

File: api/v1alpha3/cloudstackcluster_webhook.go (hunk 1 above)
Comment (maxdrib) [reply api/v1alpha3/cloudstackcluster_webhook.go:65]: isn't domain itself sufficient for performing all the necessary operations? Why do we also need account? Also is this an account ID? Or the account name?

Comment (rejoshed) [reply api/v1alpha3/cloudstackcluster_webhook.go:65]: Sorry, I worded it wrong.

If account is set, domain must be set too.

https://cloudstack.apache.org/api/apidocs-4.16/apis/deployVirtualMachine.html

It's a restriction imposed by the cloudstack api itself.

> 

account | an optional account for the virtual machine. Must be used with domainId. | false
-- | -- | --

Comment (maxdrib) [reply api/v1alpha3/cloudstackcluster_webhook.go:65]: right so wouldn't this check fail if only domain is provided, even though that's an acceptable situation?

**Pull request merged**